File: src/bindings/zig/corsa_virtual_document.zig
const std = @import("std");
const utils = @import("corsa_utils.zig");

pub const c = utils.c;

pub const VirtualDocument = struct {
    handle: ?*c.CorsaVirtualDocument,

    pub fn init(uri: []const u8, language_id: []const u8, text: []const u8) !VirtualDocument {
        return .{ .handle = c.corsa_virtual_document_new(utils.toRef(uri), utils.toRef(language_id), utils.toRef(text)) orelse return error.CorsaFfiError };
    }

    pub fn untitled(path: []const u8, language_id: []const u8, text: []const u8) !VirtualDocument {
        return .{ .handle = c.corsa_virtual_document_untitled(utils.toRef(path), utils.toRef(language_id), utils.toRef(text)) orelse return error.CorsaFfiError };
    }

    pub fn inMemory(authority: []const u8, path: []const u8, language_id: []const u8, text: []const u8) !VirtualDocument {
        return .{ .handle = c.corsa_virtual_document_in_memory(utils.toRef(authority), utils.toRef(path), utils.toRef(language_id), utils.toRef(text)) orelse return error.CorsaFfiError };
    }

    pub fn deinit(self: *VirtualDocument) void {
        if (self.handle) |handle| c.corsa_virtual_document_free(handle);
        self.handle = null;
    }

    pub fn uri(self: VirtualDocument, allocator: std.mem.Allocator) ![]u8 {
        return utils.takeString(allocator, c.corsa_virtual_document_uri(self.handle));
    }

    pub fn languageId(self: VirtualDocument, allocator: std.mem.Allocator) ![]u8 {
        return utils.takeString(allocator, c.corsa_virtual_document_language_id(self.handle));
    }

    pub fn text(self: VirtualDocument, allocator: std.mem.Allocator) ![]u8 {
        return utils.takeString(allocator, c.corsa_virtual_document_text(self.handle));
    }

    pub fn key(self: VirtualDocument, allocator: std.mem.Allocator) ![]u8 {
        return utils.takeString(allocator, c.corsa_virtual_document_key(self.handle));
    }

    pub fn version(self: VirtualDocument) i32 {
        return c.corsa_virtual_document_version(self.handle);
    }

    pub fn replace(self: VirtualDocument, text: []const u8) !void {
        if (!c.corsa_virtual_document_replace(self.handle, utils.toRef(text))) return error.CorsaFfiError;
    }

    pub fn splice(
        self: VirtualDocument,
        start_line: u32,
        start_character: u32,
        end_line: u32,
        end_character: u32,
        text: []const u8,
    ) !void {
        if (!c.corsa_virtual_document_splice(
            self.handle,
            start_line,
            start_character,
            end_line,
            end_character,
            utils.toRef(text),
        )) return error.CorsaFfiError;
    }
};

pub fn takeLastError(allocator: std.mem.Allocator) ![]u8 {
    return utils.takeString(allocator, c.corsa_error_message_take());
}

File: src/bindings/swift/CorsaUtils/Sources/CorsaUtils/CorsaVirtualDocument.swift
public enum CorsaFfiError: Error {
    case message(String)
}

@_silgen_name("corsa_error_message_take")
func takeErrorMessageNative() -> CorsaString

@_silgen_name("corsa_virtual_document_new")
private func createVirtualDocumentNative(_ uri: CorsaStrRef, _ languageID: CorsaStrRef, _ text: CorsaStrRef) -> UnsafeMutableRawPointer?

@_silgen_name("corsa_virtual_document_untitled")
private func createUntitledVirtualDocumentNative(_ path: CorsaStrRef, _ languageID: CorsaStrRef, _ text: CorsaStrRef) -> UnsafeMutableRawPointer?

@_silgen_name("corsa_virtual_document_in_memory")
private func createInMemoryVirtualDocumentNative(
    _ authority: CorsaStrRef,
    _ path: CorsaStrRef,
    _ languageID: CorsaStrRef,
    _ text: CorsaStrRef
) -> UnsafeMutableRawPointer?

@_silgen_name("corsa_virtual_document_uri")
private func virtualDocumentUriNative(_ value: UnsafeMutableRawPointer?) -> CorsaString

@_silgen_name("corsa_virtual_document_language_id")
private func virtualDocumentLanguageIdNative(_ value: UnsafeMutableRawPointer?) -> CorsaString

@_silgen_name("corsa_virtual_document_text")
private func virtualDocumentTextNative(_ value: UnsafeMutableRawPointer?) -> CorsaString

@_silgen_name("corsa_virtual_document_key")
private func virtualDocumentKeyNative(_ value: UnsafeMutableRawPointer?) -> CorsaString

@_silgen_name("corsa_virtual_document_version")
private func virtualDocumentVersionNative(_ value: UnsafeMutableRawPointer?) -> Int32

@_silgen_name("corsa_virtual_document_replace")
private func virtualDocumentReplaceNative(_ value: UnsafeMutableRawPointer?, _ text: CorsaStrRef) -> Bool

@_silgen_name("corsa_virtual_document_splice")
private func virtualDocumentSpliceNative(
    _ value: UnsafeMutableRawPointer?,
    _ startLine: UInt32,
    _ startCharacter: UInt32,
    _ endLine: UInt32,
    _ endCharacter: UInt32,
    _ text: CorsaStrRef
) -> Bool

@_silgen_name("corsa_virtual_document_free")
private func freeVirtualDocumentNative(_ value: UnsafeMutableRawPointer?)

public final class CorsaVirtualDocument {
    private var handle: UnsafeMutableRawPointer?

    private init(handle: UnsafeMutableRawPointer) {
        self.handle = handle
    }

    public init(uri: String, languageID: String, text: String) throws {
        self.handle = try CorsaVirtualDocument.create(uri: uri, languageID: languageID, text: text)
    }

    public static func untitled(path: String, languageID: String, text: String) throws -> CorsaVirtualDocument {
        try CorsaVirtualDocument(handle: withThreeRefs(path, languageID, text, createUntitledVirtualDocumentNative))
    }

    public static func inMemory(authority: String, path: String, languageID: String, text: String) throws -> CorsaVirtualDocument {
        try CorsaVirtualDocument(handle: withFourRefs(authority, path, languageID, text, createInMemoryVirtualDocumentNative))
    }

    deinit {
        freeVirtualDocumentNative(handle)
    }

    public var uri: String { takeString(virtualDocumentUriNative(handle)) }
    public var languageID: String { takeString(virtualDocumentLanguageIdNative(handle)) }
    public var text: String { takeString(virtualDocumentTextNative(handle)) }
    public var key: String { takeString(virtualDocumentKeyNative(handle)) }
    public var version: Int32 { virtualDocumentVersionNative(handle) }

    public func replace(_ text: String) throws {
        let ok = withStrRef(text) { virtualDocumentReplaceNative(handle, $0) }
        if !ok { throw ffiError() }
    }

    public func splice(
        startLine: UInt32,
        startCharacter: UInt32,
        endLine: UInt32,
        endCharacter: UInt32,
        text: String
    ) throws {
        let ok = withStrRef(text) {
            virtualDocumentSpliceNative(handle, startLine, startCharacter, endLine, endCharacter, $0)
        }
        if !ok { throw ffiError() }
    }

    private static func create(uri: String, languageID: String, text: String) throws -> UnsafeMutableRawPointer {
        try withThreeRefs(uri, languageID, text, createVirtualDocumentNative)
    }
}

private func withThreeRefs(
    _ first: String,
    _ second: String,
    _ third: String,
    _ body: (CorsaStrRef, CorsaStrRef, CorsaStrRef) -> UnsafeMutableRawPointer?
) throws -> UnsafeMutableRawPointer {
    let refs = BorrowedRefs([first, second, third])
    return try refs.refs.withUnsafeBufferPointer {
        guard let handle = body($0[0], $0[1], $0[2]) else { throw ffiError() }
        return handle
    }
}

private func withFourRefs(
    _ first: String,
    _ second: String,
    _ third: String,
    _ fourth: String,
    _ body: (CorsaStrRef, CorsaStrRef, CorsaStrRef, CorsaStrRef) -> UnsafeMutableRawPointer?
) throws -> UnsafeMutableRawPointer {
    let refs = BorrowedRefs([first, second, third, fourth])
    return try refs.refs.withUnsafeBufferPointer {
        guard let handle = body($0[0], $0[1], $0[2], $0[3]) else { throw ffiError() }
        return handle
    }
}

func ffiError() -> CorsaFfiError {
    let message = takeString(takeErrorMessageNative())
    return .message(message.isEmpty ? "corsa ffi error" : message)
}
